controllers/fluxcd/git-repository-controller.go

```diff
@@ -0,0 +1,170 @@
+		err = client.IgnoreNotFound(err)
+		return
+	}
+
+	err = r.reconcileFluxGitRepo(repo)
+	return
+}
+
+func (r *GitRepositoryReconciler) reconcileFluxGitRepo(repo *v1alpha3.GitRepository) (err error) {
+	ctx := context.Background()
+	fluxGitRepo := createBareFluxGitRepoObject()
+	// FluxGitRepo's namespace = v1alpha3.GitRepository's namespace
+	// FluxGitRepo's name = "fluxcd-" + v1alpha3.GitRepository's name
+	ns, name := repo.GetNamespace(), getFluxRepoName(repo.GetName())
+
+	if !repo.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err = r.Get(ctx, types.NamespacedName{Namespace: ns, Name: name}, fluxGitRepo); err != nil {
+			err = client.IgnoreNotFound(err)
+		} else {
+			r.log.Info("delete FluxCDGitRepository", "name", fluxGitRepo.GetName())
+			err = r.Delete(ctx, fluxGitRepo)
+		}
+
+		if err == nil {
+			k8sutil.RemoveFinalizer(&repo.ObjectMeta, v1alpha3.GitRepoFinalizerName)
+			err = r.Update(ctx, repo)
+		}
+		return
+	}
+	if k8sutil.AddFinalizer(&repo.ObjectMeta, v1alpha3.GitRepoFinalizerName) {
+		err = r.Update(ctx, repo)
+		if err != nil {
+			return
+		}
+	}
+
+	if err = r.Get(ctx, types.NamespacedName{Namespace: ns, Name: name}, fluxGitRepo); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return
+		}
+		// flux git repo did not existed
+		// create
+		newFluxGitRepo := createUnstructuredFluxGitRepo(repo)
+		if err = r.Create(ctx, newFluxGitRepo); err != nil {
+			r.recorder.Eventf(newFluxGitRepo, v1.EventTypeWarning, "FailedWithFluxCD",
+				"failed to create FluxCDGitRepository, error is: %v", err)
+		}
+	} else {
+		// flux git repo existed
+		// update
+		newFluxGitRepo := createUnstructuredFluxGitRepo(repo)
+		fluxGitRepo.Object["spec"] = newFluxGitRepo.Object["spec"]
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+			latestGitRepo := createBareFluxGitRepoObject()
+			if err = r.Get(ctx, types.NamespacedName{
+				Namespace: fluxGitRepo.GetNamespace(),
+				Name:      fluxGitRepo.GetName(),
+			}, latestGitRepo); err != nil {
+				return
+			}
+
+			fluxGitRepo.SetResourceVersion(latestGitRepo.GetResourceVersion())
+			r.log.Info("update FluxCDGitRepository", "name", fluxGitRepo.GetName())
+			err = r.Update(ctx, fluxGitRepo)
+			return
+		})
+
+	}
+	return
+}
+
+func createUnstructuredFluxGitRepo(repo *v1alpha3.GitRepository) *unstructured.Unstructured {
+	newFluxGitRepo := createBareFluxGitRepoObject()
+	newFluxGitRepo.SetNamespace(repo.GetNamespace())
+	newFluxGitRepo.SetName(getFluxRepoName(repo.GetName()))
+
+	// set url
+	_ = unstructured.SetNestedField(newFluxGitRepo.Object, repo.Spec.URL, "spec", "url")
+	// set interval
+	_ = unstructured.SetNestedField(newFluxGitRepo.Object, "1m", "spec", "interval")
+	// set secretRef
+	if repo.Spec.Secret != nil && repo.Spec.Secret.Name != "" {
+		_ = unstructured.SetNestedField(newFluxGitRepo.Object, repo.Spec.Secret.Name, "spec", "secretRef", "name")
+	}
+
+	//TODO: set other optional fields, like Ref ...
+
```

Comment:
Should we add some annotations to the created object such as `managed-by: devops.kubesphere.io`  to indicate that this resource is managed by devops

Reply:
Good idea. But it looks like a label, for example:

```yaml
labels:
  app.kubernetes.io/managed-by: Helm
```

Reply:
Great. we can use labels like this: 
```yaml
labels:
  app.kubernetes.io/managed-by: devops.kubesphere.io
```